pyrsistent/__init__.pyi

```diff
@@ -18,77 +19,127 @@ from typing import Type
+# see class_test.py for use cases
+Invariant = Tuple[bool, Optional[Union[str, Callable[[], str]]]]
+
+@overload
 def field(
-    type: Union[Type[T], Sequence[Type[T]]] = (),
-    invariant: Callable[[Any], Tuple[bool, Optional[str]]] = lambda _: (True, None),
+    type: Type[T] = ...,
+    invariant: Callable[[Any], Union[Invariant, Iterable[Invariant]]] = lambda _: (True, None),
     initial: Any = object(),
     mandatory: bool = False,
     factory: Callable[[Any], T] = lambda x: x,
     serializer: Callable[[Any, T], Any] = lambda _, value: value,
 ) -> T: ...
+# The actual return value (_PField) is irrelevant after a PRecord has been instantiated,
+# see https://github.com/tobgu/pyrsistent/blob/master/pyrsistent/_precord.py#L10
+@overload
+def field(
+    type: Any = ...,
+    invariant: Callable[[Any], Union[Invariant, Iterable[Invariant]]] = lambda _: (True, None),
+    initial: Any = object(),
+    mandatory: bool = False,
+    factory: Callable[[Any], Any] = lambda x: x,
+    serializer: Callable[[Any, Any], Any] = lambda _, value: value,
+) -> Any: ...
 
+# Use precise types for the simplest use cases, but fall back to Any for
+# everything else. See record_test.py for the wide range of possible types for
+# item_type
+@overload
 def pset_field(
     item_type: Type[T],
     optional: bool = False,
-    initial: Any = (),
+    initial: Iterable[T] = ...,
 ) -> PSet[T]: ...
+@overload
+def pset_field(
+    item_type: Any,
+    optional: bool = False,
+    initial: Any = (),
+) -> PSet[Any]: ...
 
+@overload
 def pmap_field(
     key_type: Type[KT],
     value_type: Type[VT],
     optional: bool = False,
     invariant: Callable[[Any], Tuple[bool, Optional[str]]] = lambda _: (True, None),
 ) -> PMap[KT, VT]: ...
+@overload
+def pmap_field(
+    key_type: Any,
+    value_type: Any,
+    optional: bool = False,
+    invariant: Callable[[Any], Tuple[bool, Optional[str]]] = lambda _: (True, None),
+) -> PMap[Any, Any]: ...
 
+@overload
 def pvector_field(
     item_type: Type[T],
     optional: bool = False,
-    initial: Any = (),
+    initial: Iterable[T] = ...,
 ) -> PVector[T]: ...
+@overload
+def pvector_field(
+    item_type: Any,
+    optional: bool = False,
+    initial: Any = (),
+) -> PVector[Any]: ...
 
 def pbag(elements: Iterable[T]) -> PBag[T]: ...
-def b(*elements: Iterable[T]) -> PBag[T]: ...
+def b(*elements: T) -> PBag[T]: ...
 
 def plist(iterable: Iterable[T] = (), reverse: bool = False) -> PList[T]: ...
-def l(*elements: Iterable[T]) -> PList[T]: ...
+def l(*elements: T) -> PList[T]: ...
 
-def pdeque(iterable: Iterable[T], maxlen: Optional[int] = None) -> PDeque[T]: ...
-def dq(*iterable: Iterable[T]) -> PDeque[T]: ...
+def pdeque(iterable: Optional[Iterable[T]] = None, maxlen: Optional[int] = None) -> PDeque[T]: ...
+def dq(*iterable: T) -> PDeque[T]: ...
 
-def optional(*typs: Iterable[type]) -> Iterable[type]: ...
+def optional(*typs: Union[str, Type]) -> Tuple[Type, ...]: ...
```

Comment:
Looks like a stringy type will still be stringy after calling optional?

Reply:
Yeah seems so. This should be good place for overload decorator, i'll push a new version which should be accurate when using `optional` with single argument.